File: src/javascript/algorithms/strings/lastSubstring/lastSubstring.js
// Two-pointer string comparison  TC: O(n) SC: O(1)

function lastSubstring(s) {
    const n = s.length;
    let i = 0;   // current best suffix start
    let j = 1;   // challenger suffix start
    let k = 0;   // comparison offset

    while (j + k < n) {
        if (s[i + k] === s[j + k]) {
            k++;
        } else if (s[i + k] < s[j + k]) {   // challenger is better
            i = Math.max(i + k + 1, j);
            j = i + 1;
            k = 0;
        } else {                              // current best is better
            j = j + k + 1;
            k = 0;
        }
    }

    return s.slice(i);
}

function runTests() {
    const tests = [
        { s: "abab",     expected: "bab" },
        { s: "leetcode", expected: "tcode" },
        { s: "cacacb",   expected: "cb" },
        { s: "zz",       expected: "zz" },
        { s: "abcde",    expected: "e" },
    ];

    tests.forEach(({ s, expected }) => {
        const result = lastSubstring(s);
        const status = result === expected ? "PASS" : "FAIL";
        console.log(`[${status}] result=${result} expected=${expected}`);
    });
}

runTests();

File: src/python/algorithms/strings/lastSubstring/last_substring.py
# Two-pointer string comparison  TC: O(n) SC: O(1)

def last_substring(s):
    n = len(s)
    i = 0   # current best suffix start
    j = 1   # challenger suffix start
    k = 0   # comparison offset

    while j + k < n:
        if s[i + k] == s[j + k]:
            k += 1
        elif s[i + k] < s[j + k]:   # challenger is better
            i = max(i + k + 1, j)
            j = i + 1
            k = 0
        else:                        # current best is better
            j = j + k + 1
            k = 0

    return s[i:]


if __name__ == "__main__":
    test_cases = [
        ("abab",     "bab"),
        ("leetcode", "tcode"),
        ("cacacb",   "cb"),
        ("zz",       "zz"),
        ("abcde",    "e"),
    ]

    for s, expected in test_cases:
        result = last_substring(s)
        status = "PASS" if result == expected else "FAIL"
        print(f"[{status}] result={result} expected={expected}")
